File: pilot/src/api/hypothesis.rs
use axum::body::Body;
use axum::extract::{Path, Query, State};
use axum::http::{header, StatusCode};
use axum::response::IntoResponse;
use axum::Json;
use uuid::Uuid;

use crate::adapter::grpc_client::GrpcAdapterClient;
use crate::app_state::AppState;
use crate::engine::executor::ExecutionConfig;
use crate::types::HypothesisStatus;

use super::error::ApiError;
use super::models::*;

// --- CRUD ---

pub async fn create(
    State(state): State<AppState>,
    Json(req): Json<CreateHypothesisRequest>,
) -> Result<impl IntoResponse, ApiError> {
    if state.sqlite.get_hypothesis_by_name(&req.name).await?.is_some() {
        return Err(ApiError::Conflict(format!("hypothesis '{}' already exists", req.name)));
    }
    if state.sqlite.get_generator(&req.generator).await?.is_none() {
        return Err(ApiError::BadRequest(format!("unknown generator: {}", req.generator)));
    }
    if let Some(adapter) = &req.adapter {
        if state.sqlite.get_adapter_by_name(adapter).await?.is_none() {
            return Err(ApiError::BadRequest(format!("unknown adapter: {adapter}")));
        }
    }

    let id = format!("hyp-{}", Uuid::now_v7());
    let tolerance_json = req.tolerance.as_ref().map(|t| serde_json::to_string(t)).transpose()
        .map_err(|e| ApiError::BadRequest(e.to_string()))?;
    let key_space_json = serde_json::to_string(&req.key_space.unwrap_or_default())
        .map_err(|e| ApiError::BadRequest(e.to_string()))?;

    let hypothesis = state.sqlite.create_hypothesis(
        &id, &req.name, &req.generator,
        req.adapter.as_deref(), req.adapter_addr.as_deref(), req.duration.as_deref(),
        tolerance_json.as_deref(), &req.checkpoint_every, &key_space_json, "{}",
    ).await?;

    state.files.create_hypothesis_dir(&id)?;
    state.managers.create_from_hypothesis(&id, &req.generator, tolerance_json).await;

    Ok((StatusCode::CREATED, Json(to_response(&hypothesis))))
}

pub async fn list(State(state): State<AppState>) -> Result<impl IntoResponse, ApiError> {
    let items: Vec<HypothesisResponse> = state.sqlite.list_hypotheses().await?.iter().map(to_response).collect();
    Ok(Json(HypothesisListResponse { items }))
}

pub async fn get_one(State(state): State<AppState>, Path(id): Path<String>) -> Result<impl IntoResponse, ApiError> {
    let h = state.sqlite.get_hypothesis(&id).await?.ok_or_else(|| ApiError::NotFound(format!("hypothesis {id} not found")))?;
    Ok(Json(to_response(&h)))
}

pub async fn update(
    State(state): State<AppState>,
    Path(id): Path<String>,
    Json(req): Json<CreateHypothesisRequest>,
) -> Result<impl IntoResponse, ApiError> {
    state.sqlite.get_hypothesis(&id).await?
        .ok_or_else(|| ApiError::NotFound(format!("hypothesis {id} not found")))?;

    if state.sqlite.get_generator(&req.generator).await?.is_none() {
        return Err(ApiError::BadRequest(format!("unknown generator: {}", req.generator)));
    }

    let tolerance_json = req.tolerance.as_ref().map(|t| serde_json::to_string(t)).transpose()
        .map_err(|e| ApiError::BadRequest(e.to_string()))?;

    let hypothesis = state.sqlite.update_hypothesis(
        &id, &req.name, &req.generator,
        req.adapter.as_deref(), req.adapter_addr.as_deref(),
        req.duration.as_deref(), tolerance_json.as_deref(),
    ).await?.ok_or_else(|| ApiError::Internal(anyhow::anyhow!("update failed")))?;

    Ok(Json(to_response(&hypothesis)))
}

pub async fn delete(State(state): State<AppState>, Path(id): Path<String>) -> Result<impl IntoResponse, ApiError> {
    if let Some(manager) = state.managers.get(&id).await {
        let mut mgr = manager.lock().await;
        if mgr.is_running() { mgr.stop_run().await?; }
    }
    state.managers.remove(&id).await;
    state.sqlite.delete_hypothesis(&id).await?;
    let _ = state.files.delete_hypothesis_dir(&id);

    // Cleanup adapter data in background (don't block the response)
    let adapters = state.sqlite.list_adapters_all().await.unwrap_or_default();
    let key_prefix = format!("/{id}/");
    tokio::spawn(async move {
        for adapter in &adapters {
            let addr = format!("http://adapter-{}:9090", adapter.name);
            let _ = GrpcAdapterClient::cleanup_prefix(&addr, &key_prefix).await;
        }
    });

    Ok(StatusCode::NO_CONTENT)
}

// --- Run Control ---

pub async fn start_run(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<impl IntoResponse, ApiError> {
    let hypothesis = state.sqlite.get_hypothesis(&id).await?
        .ok_or_else(|| ApiError::NotFound(format!("hypothesis {id} not found")))?;

    let manager = state.managers.get(&id).await
        .ok_or_else(|| ApiError::NotFound(format!("manager for {id} not found")))?;

    let duration_secs = hypothesis.duration.as_ref().and_then(|d| parse_duration(d));

    let checkpoint_interval_secs = parse_duration(&hypothesis.checkpoint_every).unwrap_or(600);

    let config = ExecutionConfig {
        checkpoint_interval_secs,
        duration_secs,
        ..ExecutionConfig::default()
    };

    let adapter = if let Some(name) = &hypothesis.adapter {
        Some(state.sqlite.get_adapter_by_name(name).await?
            .ok_or_else(|| ApiError::NotFound(format!("adapter '{name}' not found")))?)
    } else { None };

    let mut gen_params = crate::generator::GenerateParams {
        generator: hypothesis.generator.clone(),
        ops: usize::MAX,
        ..crate::generator::GenerateParams::default()
    };
    // Apply stored key_space config (e.g. count)
    if let Ok(ks) = serde_json::from_str::<crate::generator::KeySpaceConfig>(&hypothesis.key_space) {
        gen_params.key_space = ks;
    }
    gen_params.key_space.prefix = format!("/{id}/");

    let mut mgr = manager.lock().await;
    if mgr.is_running() {
        return Err(ApiError::Conflict("hypothesis is already running".to_string()));
    }

    let (run_id, _) = mgr.start_run(config, gen_params, adapter, hypothesis.adapter_addr.clone()).await?;

    Ok((StatusCode::ACCEPTED, Json(StartRunResponse {
        run_id, hypothesis_id: id, status: HypothesisStatus::Running.to_string(),
    })))
}

pub async fn stop_run(State(state): State<AppState>, Path(id): Path<String>) -> Result<impl IntoResponse, ApiError> {
    let manager = state.managers.get(&id).await
        .ok_or_else(|| ApiError::NotFound(format!("hypothesis {id} not found")))?;
    manager.lock().await.stop_run().await?;
    Ok(StatusCode::NO_CONTENT)
}

// --- Observability ---

pub async fn status(State(state): State<AppState>, Path(id): Path<String>) -> Result<impl IntoResponse, ApiError> {
    let h = state.sqlite.get_hypothesis(&id).await?.ok_or_else(|| ApiError::NotFound(format!("hypothesis {id} not found")))?;
    let (run_id, progress) = if let Some(mgr_arc) = state.managers.get(&id).await {
        let mgr = mgr_arc.lock().await;
        (mgr.run_id().map(|s| s.to_string()), if let Some(p) = mgr.progress() { Some(p.read().await.clone()) } else { None })
    } else { (None, None) };
    Ok(Json(StatusResponse { hypothesis_id: id, status: h.status, run_id, progress }))
}

pub async fn events(State(state): State<AppState>, Path(id): Path<String>, Query(query): Query<EventsQuery>) -> Result<impl IntoResponse, ApiError> {
    state.sqlite.get_hypothesis(&id).await?.ok_or_else(|| ApiError::NotFound(format!("hypothesis {id} not found")))?;
    let events = state.files.read_merged_events(&id, query.run_id.as_deref(), query.event_type.as_deref(), query.since.as_deref())?;
    let tail = if let Some(last) = query.last {
        if events.len() > last { &events[events.len() - last..] } else { &events }
    } else {
        &events
    };
    let mut body = String::new();
    for event in tail { body.push_str(&serde_json::to_string(event).unwrap_or_default()); body.push('\n'); }
    Ok(([(header::CONTENT_TYPE, "application/x-ndjson")], body))
}

pub async fn results(State(state): State<AppState>, Path(id): Path<String>) -> Result<impl IntoResponse, ApiError> {
    state.sqlite.get_hypothesis(&id).await?.ok_or_else(|| ApiError::NotFound(format!("hypothesis {id} not found")))?;
    let results = state.sqlite.get_results(&id).await?;
    let items: Vec<serde_json::Value> = results.iter().map(|r| serde_json::json!({
        "id": r.id, "run_id": r.run_id, "total_batches": r.total_batches,
        "total_checkpoints": r.total_checkpoints, "passed_checkpoints": r.passed_checkpoints,
        "failed_checkpoints": r.failed_checkpoints, "total_response_ops": r.total_response_ops,
        "safety_violations": r.safety_violations, "stop_reason": r.stop_reason,
        "started_at": r.started_at, "finished_at": r.finished_at,
    })).collect();
    Ok(Json(serde_json::json!({ "items": items })))
}

pub async fn delete_result(
    State(state): State<AppState>,
    Path((id, result_id)): Path<(String, String)>,
) -> Result<impl IntoResponse, ApiError> {
    state.sqlite.get_hypothesis(&id).await?
        .ok_or_else(|| ApiError::NotFound(format!("hypothesis {id} not found")))?;
    let deleted = state.sqlite.delete_result(&result_id).await?;
    if !deleted {
        return Err(ApiError::NotFound(format!("result {result_id} not found")));
    }
    Ok(StatusCode::NO_CONTENT)
}

pub async fn bundle(State(state): State<AppState>, Path(id): Path<String>, Query(query): Query<BundleQuery>) -> Result<impl IntoResponse, ApiError> {
    state.sqlite.get_hypothesis(&id).await?.ok_or_else(|| ApiError::NotFound(format!("hypothesis {id} not found")))?;
    let data = state.files.create_bundle(&id, query.run_id.as_deref())?;
    let disposition = format!("attachment; filename=\"hypothesis-{id}.zip\"");
    Ok((StatusCode::OK, [(header::CONTENT_TYPE.as_str(), "application/zip".to_string()), (header::CONTENT_DISPOSITION.as_str(), disposition)], Body::from(data)))
}

// --- Helpers ---

fn parse_duration(s: &str) -> Option<u64> {
    let s = s.trim();
    if s.ends_with('s') { s[..s.len()-1].parse().ok() }
    else if s.ends_with('m') { s[..s.len()-1].parse::<u64>().ok().map(|m| m * 60) }
    else if s.ends_with('h') { s[..s.len()-1].parse::<u64>().ok().map(|h| h * 3600) }
    else { s.parse().ok() }
}

fn to_response(h: &crate::storage::sqlite::Hypothesis) -> HypothesisResponse {
    HypothesisResponse {
        id: h.id.clone(),
        name: h.name.clone(),
        generator: h.generator.clone(),
        adapter: h.adapter.clone(),
        adapter_addr: h.adapter_addr.clone(),
        duration: h.duration.clone(),
        checkpoint_every: h.checkpoint_every.clone(),
        tolerance: h.tolerance.as_ref().and_then(|t| serde_json::from_str(t).ok()),
        status: h.status.clone(),
        created_at: h.created_at.clone(),
        last_run_at: h.last_run_at.clone(),
    }
}

File: pilot/endpoints/hypothesis.go
package endpoints

import (
	"context"
	"log/slog"

	"github.com/regret-io/regret/pilot-go/ext"
	"encoding/json"
	"fmt"
	"math"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"

	"github.com/regret-io/regret/pilot-go/adapter"
	"github.com/regret-io/regret/pilot-go/database"
	"github.com/regret-io/regret/pilot-go/engine"
	"github.com/regret-io/regret/pilot-go/generator"
)

type hypothesisHandlers struct {
	state *AppState
}

// Create handles POST /api/hypothesis.
func (h *hypothesisHandlers) Create(w http.ResponseWriter, r *http.Request) {
	var req CreateHypothesisRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		WriteError(w, BadRequest("invalid JSON: "+err.Error()))
		return
	}
	if req.CheckpointEvery == "" {
		req.CheckpointEvery = "10m"
	}

	ctx := r.Context()

	// Check name uniqueness
	existing, _ := h.state.Sqlite.GetHypothesisByName(ctx, req.Name)
	if existing != nil {
		WriteError(w, Conflict(fmt.Sprintf("hypothesis '%s' already exists", req.Name)))
		return
	}

	// Validate generator
	gen, err := h.state.Sqlite.GetGenerator(ctx, req.Generator)
	if err != nil || gen == nil {
		WriteError(w, BadRequest(fmt.Sprintf("unknown generator: %s", req.Generator)))
		return
	}

	// Validate adapter
	if req.Adapter != nil {
		a, err := h.state.Sqlite.GetAdapterByName(ctx, *req.Adapter)
		if err != nil || a == nil {
			WriteError(w, BadRequest(fmt.Sprintf("unknown adapter: %s", *req.Adapter)))
			return
		}
	}

	id := fmt.Sprintf("hyp-%d", ext.NowUnixMilli())

	var toleranceJSON string
	if req.Tolerance != nil {
		toleranceJSON = string(*req.Tolerance)
	}

	var keySpaceJSON string
	if req.KeySpace != nil {
		keySpaceJSON = string(*req.KeySpace)
	} else {
		keySpaceJSON = "{}"
	}

	adapterStr := ""
	if req.Adapter != nil {
		adapterStr = *req.Adapter
	}
	adapterAddrStr := ""
	if req.AdapterAddr != nil {
		adapterAddrStr = *req.AdapterAddr
	}
	durationStr := ""
	if req.Duration != nil {
		durationStr = *req.Duration
	}

	hyp, err := h.state.Sqlite.CreateHypothesis(ctx, id, req.Name, req.Generator,
		adapterStr, adapterAddrStr, durationStr, toleranceJSON, req.CheckpointEvery, keySpaceJSON, "{}")
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}

	_ = h.state.Files.CreateHypothesisDir(id)

	tolPtr := &toleranceJSON
	if toleranceJSON == "" {
		tolPtr = nil
	}
	h.state.Managers.CreateFromHypothesis(id, req.Generator, tolPtr)

	WriteJSON(w, http.StatusCreated, toHypothesisResponse(hyp))
}

// List handles GET /api/hypothesis.
func (h *hypothesisHandlers) List(w http.ResponseWriter, r *http.Request) {
	items, err := h.state.Sqlite.ListHypotheses(r.Context())
	if err != nil {
		slog.Error("ListHypotheses failed", "error", err)
		WriteError(w, InternalError(err))
		return
	}
	slog.Debug("ListHypotheses", "count", len(items))

	resp := make([]HypothesisResponse, 0, len(items))
	for i := range items {
		resp = append(resp, toHypothesisResponse(&items[i]))
	}
	WriteJSON(w, http.StatusOK, HypothesisListResponse{Items: resp})
}

// GetOne handles GET /api/hypothesis/{id}.
func (h *hypothesisHandlers) GetOne(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	hyp, err := h.state.Sqlite.GetHypothesis(r.Context(), id)
	if err != nil {
		WriteError(w, NotFound(fmt.Sprintf("hypothesis %s not found", id)))
		return
	}
	WriteJSON(w, http.StatusOK, toHypothesisResponse(hyp))
}

// Update handles PUT /api/hypothesis/{id}.
func (h *hypothesisHandlers) Update(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	ctx := r.Context()

	_, err := h.state.Sqlite.GetHypothesis(ctx, id)
	if err != nil {
		WriteError(w, NotFound(fmt.Sprintf("hypothesis %s not found", id)))
		return
	}

	var req CreateHypothesisRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		WriteError(w, BadRequest("invalid JSON: "+err.Error()))
		return
	}

	gen, err := h.state.Sqlite.GetGenerator(ctx, req.Generator)
	if err != nil || gen == nil {
		WriteError(w, BadRequest(fmt.Sprintf("unknown generator: %s", req.Generator)))
		return
	}

	var toleranceJSON string
	if req.Tolerance != nil {
		toleranceJSON = string(*req.Tolerance)
	}

	adapterStr := ""
	if req.Adapter != nil {
		adapterStr = *req.Adapter
	}
	adapterAddrStr := ""
	if req.AdapterAddr != nil {
		adapterAddrStr = *req.AdapterAddr
	}
	durationStr := ""
	if req.Duration != nil {
		durationStr = *req.Duration
	}
	checkpointEvery := req.CheckpointEvery
	if checkpointEvery == "" {
		checkpointEvery = "10m"
	}
	keySpaceJSON := "{}"
	if req.KeySpace != nil {
		keySpaceJSON = string(*req.KeySpace)
	}

	if err := h.state.Sqlite.UpdateHypothesis(ctx, id, req.Name, req.Generator,
		adapterStr, adapterAddrStr, durationStr, toleranceJSON, checkpointEvery, keySpaceJSON); err != nil {
		WriteError(w, InternalError(err))
		return
	}

	hyp, err := h.state.Sqlite.GetHypothesis(ctx, id)
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}
	WriteJSON(w, http.StatusOK, toHypothesisResponse(hyp))
}

// Delete handles DELETE /api/hypothesis/{id}.
func (h *hypothesisHandlers) Delete(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	ctx := r.Context()

	mgr := h.state.Managers.Get(id)
	if mgr != nil {
		if mgr.IsRunning() {
			_ = mgr.StopRun()
		}
	}
	h.state.Managers.Remove(id)
	_ = h.state.Sqlite.DeleteHypothesis(ctx, id)
	_ = h.state.Files.DeleteHypothesisDir(id)

	// Best-effort adapter cleanup in background
	adapters, _ := h.state.Sqlite.ListAdapters(ctx)
	keyPrefix := fmt.Sprintf("/%s/", id)
	go func() {
		for _, a := range adapters {
			addr := fmt.Sprintf("http://adapter-%s:9090", a.Name)
			_ = adapter.CleanupPrefix(addr, keyPrefix)
		}
	}()

	w.WriteHeader(http.StatusNoContent)
}

// StartRun handles POST /api/hypothesis/{id}/run.
func (h *hypothesisHandlers) StartRun(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	ctx := r.Context()

	hyp, err := h.state.Sqlite.GetHypothesis(ctx, id)
	if err != nil {
		WriteError(w, NotFound(fmt.Sprintf("hypothesis %s not found", id)))
		return
	}

	mgr := h.state.Managers.Get(id)
	if mgr == nil {
		WriteError(w, NotFound(fmt.Sprintf("manager for %s not found", id)))
		return
	}

	if mgr.IsRunning() {
		WriteError(w, Conflict("hypothesis is already running"))
		return
	}

	durationSecs := ext.ParseDuration(hyp.Duration)
	checkpointIntervalSecs := ext.ParseDurationSecs(hyp.CheckpointEvery)
	if checkpointIntervalSecs == 0 {
		checkpointIntervalSecs = 600
	}

	config := engine.ExecutionConfig{
		BatchSize:              100,
		CheckpointIntervalSecs: checkpointIntervalSecs,
		FailFast:               true,
		DurationSecs:           durationSecs,
	}

	var adapterRecord *storageAdapterRecord
	if hyp.Adapter != nil {
		a, err := h.state.Sqlite.GetAdapterByName(ctx, *hyp.Adapter)
		if err != nil {
			WriteError(w, NotFound(fmt.Sprintf("adapter '%s' not found", *hyp.Adapter)))
			return
		}
		adapterRecord = a
	}

	genRecord, _ := h.state.Sqlite.GetGenerator(ctx, hyp.Generator)
	var rate uint32
	if genRecord != nil {
		rate = uint32(genRecord.Rate)
	}
	skipWarmup := strings.Contains(hyp.Generator, "notification")

	genParams := generator.DefaultGenerateParams()
	genParams.Generator = hyp.Generator
	genParams.Ops = math.MaxInt32
	genParams.SkipWarmup = skipWarmup
	genParams.Rate = rate

	// Apply stored key_space config (preserve defaults for zero values)
	if hyp.KeySpace != "" {
		var ks generator.KeySpaceConfig
		if json.Unmarshal([]byte(hyp.KeySpace), &ks) == nil {
			if ks.Count > 0 {
				genParams.KeySpace.Count = ks.Count
			}
			if ks.Prefix != "" {
				genParams.KeySpace.Prefix = ks.Prefix
			}
		}
	}
	genParams.KeySpace.Prefix = fmt.Sprintf("/%s/", id)

	// Connect to adapter if specified
	var adapterClient engine.AdapterClient
	if adapterRecord != nil {
		addr := ""
		if hyp.AdapterAddr != nil && *hyp.AdapterAddr != "" {
			addr = *hyp.AdapterAddr
		} else {
			addr = fmt.Sprintf("http://adapter-%s:9090", adapterRecord.Name)
		}
		client, err := adapter.Connect(addr)
		if err == nil {
			adapterClient = client
		}
	}

	runID, _, err := mgr.StartRun(context.Background(), config, genParams, adapterRecord, hyp.AdapterAddr, adapterClient)
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}

	WriteJSON(w, http.StatusAccepted, StartRunResponse{
		RunID:        runID,
		HypothesisID: id,
		Status:       "running",
	})
}

// StopRun handles DELETE /api/hypothesis/{id}/run.
func (h *hypothesisHandlers) StopRun(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	mgr := h.state.Managers.Get(id)
	if mgr == nil {
		WriteError(w, NotFound(fmt.Sprintf("hypothesis %s not found", id)))
		return
	}
	if err := mgr.StopRun(); err != nil {
		WriteError(w, InternalError(err))
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// GetStatus handles GET /api/hypothesis/{id}/status.
func (h *hypothesisHandlers) GetStatus(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	hyp, err := h.state.Sqlite.GetHypothesis(r.Context(), id)
	if err != nil {
		WriteError(w, NotFound(fmt.Sprintf("hypothesis %s not found", id)))
		return
	}

	resp := StatusResponse{
		HypothesisID: id,
		Status:       hyp.Status,
	}

	mgr := h.state.Managers.Get(id)
	if mgr != nil {
		resp.RunID = mgr.RunID()
		resp.Progress = mgr.Progress()
	}

	WriteJSON(w, http.StatusOK, resp)
}

// GetEvents handles GET /api/hypothesis/{id}/events.
func (h *hypothesisHandlers) GetEvents(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	ctx := r.Context()

	_, err := h.state.Sqlite.GetHypothesis(ctx, id)
	if err != nil {
		WriteError(w, NotFound(fmt.Sprintf("hypothesis %s not found", id)))
		return
	}

	q := r.URL.Query()
	runID := queryPtr(q, "run_id")
	eventType := queryPtr(q, "type")
	since := queryPtr(q, "since")
	last := queryIntPtr(q, "last")

	events, err := h.state.Files.ReadMergedEvents(id, runID, eventType, since, last)
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}

	w.Header().Set("Content-Type", "application/x-ndjson")
	w.WriteHeader(http.StatusOK)
	for _, event := range events {
		data, _ := json.Marshal(event)
		w.Write(data)
		w.Write([]byte("\n"))
	}
}

// GetResults handles GET /api/hypothesis/{id}/results.
func (h *hypothesisHandlers) GetResults(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	ctx := r.Context()

	_, err := h.state.Sqlite.GetHypothesis(ctx, id)
	if err != nil {
		WriteError(w, NotFound(fmt.Sprintf("hypothesis %s not found", id)))
		return
	}

	results, err := h.state.Sqlite.GetResults(ctx, id)
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}

	type resultItem struct {
		ID                string  `json:"id"`
		RunID             string  `json:"run_id"`
		TotalBatches      int64   `json:"total_batches"`
		TotalCheckpoints  int64   `json:"total_checkpoints"`
		PassedCheckpoints int64   `json:"passed_checkpoints"`
		FailedCheckpoints int64   `json:"failed_checkpoints"`
		TotalResponseOps  int64   `json:"total_response_ops"`
		SafetyViolations  int64   `json:"safety_violations"`
		StopReason        *string `json:"stop_reason,omitempty"`
		StartedAt         *string `json:"started_at,omitempty"`
		FinishedAt        *string `json:"finished_at,omitempty"`
	}

	items := make([]resultItem, 0, len(results))
	for _, r := range results {
		items = append(items, resultItem{
			ID: r.ID, RunID: r.RunID, TotalBatches: r.TotalBatches,
			TotalCheckpoints: r.TotalCheckpoints, PassedCheckpoints: r.PassedCheckpoints,
			FailedCheckpoints: r.FailedCheckpoints, TotalResponseOps: r.TotalResponseOps,
			SafetyViolations: r.SafetyViolations, StopReason: r.StopReason,
			StartedAt: r.StartedAt, FinishedAt: r.FinishedAt,
		})
	}

	WriteJSON(w, http.StatusOK, map[string]interface{}{"items": items})
}

// DeleteResult handles DELETE /api/hypothesis/{id}/results/{resultID}.
func (h *hypothesisHandlers) DeleteResult(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	resultID := chi.URLParam(r, "resultID")
	ctx := r.Context()

	_, err := h.state.Sqlite.GetHypothesis(ctx, id)
	if err != nil {
		WriteError(w, NotFound(fmt.Sprintf("hypothesis %s not found", id)))
		return
	}

	deleted, err := h.state.Sqlite.DeleteResult(ctx, resultID)
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}
	if !deleted {
		WriteError(w, NotFound(fmt.Sprintf("result %s not found", resultID)))
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// GetBundle handles GET /api/hypothesis/{id}/bundle.
func (h *hypothesisHandlers) GetBundle(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	ctx := r.Context()

	_, err := h.state.Sqlite.GetHypothesis(ctx, id)
	if err != nil {
		WriteError(w, NotFound(fmt.Sprintf("hypothesis %s not found", id)))
		return
	}

	runID := queryPtr(r.URL.Query(), "run_id")
	data, err := h.state.Files.CreateBundle(id, runID)
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}

	w.Header().Set("Content-Type", "application/zip")
	w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="hypothesis-%s.zip"`, id))
	w.WriteHeader(http.StatusOK)
	w.Write(data)
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

// storageAdapterRecord is a type alias to avoid import cycles.
type storageAdapterRecord = database.AdapterRecord

func toHypothesisResponse(hyp *database.Hypothesis) HypothesisResponse {
	resp := HypothesisResponse{
		ID:              hyp.ID,
		Name:            hyp.Name,
		Generator:       hyp.Generator,
		Adapter:         hyp.Adapter,
		AdapterAddr:     hyp.AdapterAddr,
		Duration:        hyp.Duration,
		CheckpointEvery: hyp.CheckpointEvery,
		Status:          hyp.Status,
		CreatedAt:       hyp.CreatedAt,
		LastRunAt:       hyp.LastRunAt,
	}
	if hyp.Tolerance != nil && *hyp.Tolerance != "" {
		raw := json.RawMessage(*hyp.Tolerance)
		resp.Tolerance = &raw
	}
	return resp
}

func queryPtr(q map[string][]string, key string) *string {
	if vals, ok := q[key]; ok && len(vals) > 0 && vals[0] != "" {
		return &vals[0]
	}
	return nil
}

func queryIntPtr(q map[string][]string, key string) *int {
	if vals, ok := q[key]; ok && len(vals) > 0 {
		v, err := strconv.Atoi(vals[0])
		if err == nil {
			return &v
		}
	}
	return nil
}


// Ensure time import is used.
var _ = time.Now
